### DISCUSSION TITLE
Global AppState in server functions

### DISCUSSION THREAD

**Discussion by Tibuarka** [leptos-rs/leptos#2367] [Q&A] [answered]
Just starting with Leptos and using the SSR Axum template. Coming from Axum originally, I am accustomed to using a global AppState struct for sharing database pools and such. I had a look around and saw nothing sound describing how to implement such a global app state quickly/easily. Am I missing something in the examples or the book perhaps?

**Comment by gbj** [comment]
In the current version I believe your state type is provide via context -- ie you can `use_context::<AppState>()` in a server function. 

Agreed it should be documented better.

**Reply by Tibuarka** [reply]
Ah, thanks I will test it after dinner. In regards to the documentation I unfortunately cannot give recommendations (lack of knowledge on my side).

**Reply by Tibuarka** [reply]
So I tested your possible solution with a new project like this:
```rust
//### main.rs
...
let app_state = modules::appstate::AppState{ title: "Hello World" };
...
let app = Router::new()
                 .leptos_routes(&leptos_options, routes, App)
                  .fallback(file_and_error_handler)
                  .with_state(leptos_options)
                  .with_state(app_state); <- my AppState

// ### random module file I crated for the struct - appstate.rs
// Struct for testing
#[derive(Clone, Debug)]
pub struct AppState{
    pub title: &'static str,
}

// ### app.rs
#[component]
fn HomePage() -> impl IntoView {
    // Creates a reactive value to update the button
    let (count, set_count) = create_signal(0);
    let on_click = move |_| set_count.update(|count| {
        *count += 1;
        // Calling the server function
        spawn_local(async {
            let _ = test().await;
        });
    });

    view! {
        <h1>"Welcome to Leptos!"</h1>
        <button on:click=on_click>"Click Me: " {count}</button>
    }
}

// Server Function I tested it with
#[server(Test, "/api/test")]
async fn test() -> Result<(), ServerFnError> {
    let state = use_context::<AppState>();
    println!("Ping");
    dbg!(state);
    Ok(())
}
```

I believe I'm doing something wrong, because I'm getting 'None' from the use_context() function

**Reply by gbj** [reply, marked as answer]
I don't think you can add multiple kinds of state like that, although I'm surprised Axum lets it compile. I am not an expert with Axum by any means.

But I put together a working example like this, using a single state type and the `FromRef` pattern in Axum:

`app_state.rs`
```rust
use axum::extract::FromRef;
use leptos::LeptosOptions;

#[derive(FromRef, Clone, Debug)]
pub struct AppState {
    pub title: &'static str,
    pub leptos_options: LeptosOptions,
}
```

`main.rs`
```rust
    let app_state = AppState {
        title: "Foo",
        leptos_options,
    };

    // build our application with a route
    let app = Router::new()
        .leptos_routes(&app_state, routes, App)
        .fallback(file_and_error_handler)
        .with_state(app_state);
```

**Reply by Tibuarka** [reply]
I just realized that I never got back to you (sorry for that). I ran a test with multiple states in a standalone axum environment. It compiled just fine but it could be because I instantiated the same appstate struct twice, I dont know. In regards to my leptos appstate problem I got it working with your help. I'll mark this thread as concluded. Thanks again for your help!
